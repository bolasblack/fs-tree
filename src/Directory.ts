import {
  Tree as ITree,
  File as IFile,
  Directory as IDirectory,
  FileVisitor,
} from './interfaces'
import { resolve } from 'path'

export interface ConstructOptions {
  tree: ITree
  listChildren: () => Promise<{
    files: IFile[]
    dirs: IDirectory[]
  }>
}

export class Directory implements IDirectory {
  constructor(readonly path: string, private _options: ConstructOptions) {}

  dir(name: string) {
    return this._options.tree.getDir(resolve(this.path, name))
  }

  file(name: string) {
    return this._options.tree.get(resolve(this.path, name))
  }

  async visit(visitor: FileVisitor) {
    const entries = await this._options.listChildren()

    for (const file of entries.files) {
      await visitor(file)
    }

    for (const dir of entries.dirs) {
      await dir.visit(visitor)
    }
  }
}
