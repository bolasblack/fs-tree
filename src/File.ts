import { File as IFile } from './interfaces'

export interface ConstructOptions {
  stat: () => Promise<IFile.Stats>
  content: () => Promise<Buffer>
}

export class File implements IFile {
  constructor(
    public readonly path: IFile['path'],
    private _options: ConstructOptions,
  ) {}

  get content() {
    return this._options.content()
  }

  get stats() {
    return this._options.stat()
  }
}
