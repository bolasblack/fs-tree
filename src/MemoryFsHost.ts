import { Volume } from 'memfs'
import { Union } from 'unionfs'
import fs from 'fs'
import { FsHostBase } from './FsHostBase'

export class MemoryFsHost extends FsHostBase {
  constructor(vol = Volume.fromJSON({})) {
    super(new Union().use(vol).use(fs))
  }
}
