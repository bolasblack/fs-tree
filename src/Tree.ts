import {
  Host as IHost,
  Tree as ITree,
  File as IFile,
  Directory as IDirectory,
  StatsModifyOptions,
} from './interfaces'
import { File } from './File'
import { Directory } from './Directory'
import {
  PathIsDirectoryException,
  PathIsFileException,
  FileDoesNotExistException,
  MergeConflictException,
} from './exceptions'
import {
  ActionCollector,
  CreateAction,
  OverwriteAction,
  MoveAction,
  DeleteAction,
} from './Action'
import { statsEquals } from './utils/statsEquals'
import { createSafeReadStat } from './utils/safeReadStat'

enum MergeStrategy {
  AllowOverwriteConflict = 1 << 1,
  AllowCreationConflict = 1 << 2,
  AllowDeleteConflict = 1 << 3,

  // Uses the default strategy.
  Default = 0,

  // Error out if 2 files have the same path. It is useful to have a different value than
  // Default in this case as the tooling Default might differ.
  Error = 1 << 0,

  // Only content conflicts are overwritten.
  ContentOnly = AllowOverwriteConflict,

  // Overwrite everything with the latest change.
  Overwrite = AllowOverwriteConflict +
    AllowCreationConflict +
    AllowDeleteConflict,
}

export class Tree implements ITree {
  static MergeStrategy = MergeStrategy

  private _actionCollector: ActionCollector

  private _getHost: () => Promise<IHost>

  private _hostRead = async (path: string) => {
    return (await this._getHost()).read(path)
  }

  private _hostReadStat = async (path: string) => {
    return (await this._getHost()).readStat(path)
  }

  private _safeReadStat = createSafeReadStat(this._hostReadStat)

  constructor(private _createHost: () => IHost | Promise<IHost>) {
    let hostP: Promise<IHost> | null = null

    this._getHost = () => {
      if (hostP) return hostP
      return (hostP = Promise.resolve(_createHost()))
    }

    this._actionCollector = new ActionCollector({
      readContent: this._hostRead,
      readStat: this._hostReadStat,
    })
  }

  async branch() {
    const newTree = new Tree(this._createHost)
    newTree._actionCollector = this._actionCollector.clone({
      readContent: newTree._hostRead,
      readStat: newTree._hostReadStat,
    })
    return newTree
  }

  async merge(other: ITree, strategy: MergeStrategy = MergeStrategy.Default) {
    if (other === this) return

    const actions = await other.exportActions()

    for (let action of actions) {
      switch (action.type) {
        case 'overwrite':
          await this._mergeOverwriteAction(action, strategy)
          break
        case 'create':
          await this._mergeCreateAction(action, strategy)
          break
        case 'move':
          await this._mergeMoveAction(action)
          break
        case 'delete':
          await this._mergeDeleteAction(action, strategy)
          break
      }
    }
  }

  exportActions() {
    return this._actionCollector.toActions()
  }

  async get(path: string): Promise<IFile | null> {
    if (await this._isDirectory(path)) {
      throw new PathIsDirectoryException(path)
    }

    if (!(await this._exists(path))) return null

    return this._getFile(path)
  }

  async getDir(path: string): Promise<IDirectory> {
    if (await this._isFile(path)) {
      throw new PathIsFileException(path)
    }

    return this._getDir(path)
  }

  async overwrite(
    path: string,
    content?: Buffer | string,
    stat?: StatsModifyOptions,
  ) {
    const originStat = await this._safeReadStat(path)

    if (!originStat) {
      throw new FileDoesNotExistException(path)
    }

    stat = this._createStat(originStat, stat || {})

    const host = await this._getHost()

    await host.overwrite(
      path,
      content != null ? this._ensureBuffer(content) : undefined,
      stat,
    )

    this._actionCollector.overwrite(path)
  }

  async create(
    path: string,
    content: Buffer | string,
    stat?: StatsModifyOptions,
  ) {
    const host = await this._getHost()

    await host.create(
      path,
      this._ensureBuffer(content),
      this._createStat(stat || {}),
    )

    this._actionCollector.create(path)
  }

  async delete(path: string) {
    await (await this._getHost()).delete(path)

    this._actionCollector.delete(path)
  }

  async move(from: string, to: string) {
    await (await this._getHost()).move(from, to)

    this._actionCollector.move(from, to)
  }

  private _createStat(
    ...stats: Partial<StatsModifyOptions>[]
  ): StatsModifyOptions {
    return Object.assign({}, ...stats)
  }

  private _ensureBuffer(content: Buffer | string) {
    if (typeof content === 'string') {
      return Buffer.from(content)
    }

    return content
  }

  private _getFile(path: string) {
    return new File(path, {
      content: () => this._hostRead(path),
      stat: () => this._hostReadStat(path),
    })
  }

  private _getDir(path: string): Directory {
    return new Directory(path, {
      tree: this,
      listChildren: this._listChildren(path),
    })
  }

  private _listChildren(path: string) {
    return async () => {
      const _host = await this._getHost()
      const dirInfo = await _host.readDir(path)

      return {
        dirs: dirInfo.dirs.map(this._getDir.bind(this)),
        files: dirInfo.files.map(this._getFile.bind(this)),
      }
    }
  }

  private async _isFile(path: string) {
    const stat = await this._safeReadStat(path)
    if (!stat) return false
    return stat.isFile()
  }

  private async _isDirectory(path: string) {
    const stat = await this._safeReadStat(path)
    if (!stat) return false
    return stat.isDirectory()
  }

  private async _exists(path: string) {
    return Boolean(await this._safeReadStat(path))
  }

  private async _mergeDeleteAction(
    action: DeleteAction,
    strategy: MergeStrategy,
  ) {
    const deleteConflictAllowed =
      (strategy & MergeStrategy.AllowOverwriteConflict) ==
      MergeStrategy.AllowDeleteConflict

    const { path } = action

    if (this._willDelete(path)) {
      // TODO: (schematic) This should technically check the content (e.g., hash on delete)
      //   But I have no idea how to implement it, so let's wait for schematic :P
      // Identical outcome; no action required
      return
    }

    if (!(await this._exists(path)) && !deleteConflictAllowed) {
      throw new MergeConflictException(path)
    }

    await this.delete(path)
  }

  private async _mergeMoveAction(action: MoveAction) {
    const { path, to } = action

    if (this._willDelete(path)) {
      throw new MergeConflictException(path)
    }

    if (this._willMove(path)) {
      if (this._willMoveTo(path, to)) {
        // Identical outcome; no action required
        return
      }

      // No override possible for renaming.
      throw new MergeConflictException(path)
    }

    await this.move(path, to)
  }

  private async _mergeOverwriteAction(
    action: OverwriteAction,
    strategy: MergeStrategy,
  ) {
    const overwriteConflictAllowed =
      (strategy & MergeStrategy.AllowOverwriteConflict) ==
      MergeStrategy.AllowOverwriteConflict

    const { path, content, stat } = action

    if (this._willDelete(path) && !overwriteConflictAllowed) {
      throw new MergeConflictException(path)
    }

    if (this._willOverwrite(path)) {
      const [existingContent, existingStat] = await Promise.all([
        this._hostRead(path),
        this._hostReadStat(path),
      ])

      // Ignore if content is the same (considered the same change).
      if (
        // prettier-ignore
        (!content || (existingContent && content.equals(existingContent))) &&
        (!stat || (existingStat && statsEquals(existingStat, stat)))
      ) {
        // Identical outcome; no action required
        return
      }

      if (!overwriteConflictAllowed) {
        throw new MergeConflictException(path)
      }
    }

    // We use write here as merge validation has already been done, and we want to let
    // the CordHost do its job.
    await this.overwrite(path, content, stat)
  }

  private async _mergeCreateAction(
    action: CreateAction,
    strategy: MergeStrategy,
  ) {
    const creationConflictAllowed =
      (strategy & MergeStrategy.AllowCreationConflict) ==
      MergeStrategy.AllowCreationConflict

    const { path, content, stat } = action

    if (this._willCreate(path) || this._willOverwrite(path)) {
      const [existingContent, existingStat] = await Promise.all([
        this._hostRead(path),
        this._hostReadStat(path),
      ])

      if (
        // prettier-ignore
        (existingContent && content.equals(existingContent)) &&
        (!stat || (existingStat && statsEquals(existingStat, stat)))
      ) {
        // Identical outcome; no action required
        return
      }

      if (!creationConflictAllowed) {
        throw new MergeConflictException(path)
      }

      await this.overwrite(path, content, stat)
    } else {
      await this.create(path, content, stat)
    }
  }

  protected _willCreate(path: string) {
    return this._actionCollector.willCreate(path)
  }

  protected _willOverwrite(path: string) {
    return this._actionCollector.willOverwrite(path)
  }

  protected _willDelete(path: string) {
    return this._actionCollector.willDelete(path)
  }

  protected _willMove(path: string) {
    return this._actionCollector.willMove(path)
  }

  protected _willMoveTo(path: string, to: string) {
    return this._actionCollector.willMoveTo(path, to)
  }
}
