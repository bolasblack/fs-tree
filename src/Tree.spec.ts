import fs from 'fs'
import path from 'path'
import { Volume } from 'memfs'
import { omit } from 'ramda'
import { MemoryFsHost } from './MemoryFsHost'
import { Tree } from './Tree'
import { File } from './File'
import { Directory } from './Directory'

describe('Tree', () => {
  describe('#get', () => {
    it('throw error if target is directory', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree({ [filePath]: 'content' })
      const getPromise = tree.get(path.resolve(filePath, '../'))
      await expect(getPromise).rejects.toMatchSnapshot()
    })

    it('return null instance if target not exists', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree()
      const file = await tree.get(filePath)
      expect(file).toBe(null)
    })

    it('return File instance if target exists', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree({ [filePath]: 'content' })
      const file = await tree.get(filePath)
      expect(file).toBeInstanceOf(File)
      expect(file!.path).toBe(filePath)
      expect((await file!.content).toString()).toBe('content')
    })

    it('support read file content', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree({ [filePath]: 'content' })
      const file = await tree.get(filePath)
      expect(file).toBeInstanceOf(File)
      expect((await file!.content).toString()).toBe('content')
    })

    it('support read file stat', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree({ [filePath]: 'content' })
      const file = await tree.get(filePath)
      expect(file).toBeInstanceOf(File)
      expect(await file!.stats).toMatchObject({
        mode: 0o100666,
        blksize: 4096,
        blocks: 1,
      })
    })
  })

  describe('#getDir', () => {
    it('throw error if target is file', async () => {
      const dirPath = '/some/path/to'
      const tree = createTree({ [dirPath + '/file']: 'content' })
      const getPromise = tree.getDir(dirPath + '/file')
      await expect(getPromise).rejects.toMatchSnapshot()
    })

    it('return Directory instance if target exists', async () => {
      const dirPath = '/some/path/to'
      const tree = createTree({ [dirPath + '/file']: 'content' })
      const dir = await tree.getDir(dirPath)
      expect(dir).toBeInstanceOf(Directory)
      expect(dir!.path).toBe(dirPath)
    })

    it('return Directory instance if target not exists', async () => {
      const dirPath = '/some/path/to'
      const tree = createTree()
      const dir = await tree.getDir(dirPath)
      expect(dir).toBeInstanceOf(Directory)
      expect(dir!.path).toBe(dirPath)
    })

    it('support read dir', async () => {
      const fsData = {
        '/some1': '/some1 content',
        '/some/file': '/some/file content',
        '/some/path/file': '/some/path/file content',
      }
      const tree = createTree(fsData)
      const dir = await tree.getDir('/some')
      const visitContent: { [path: string]: string } = {}
      await dir.visit(async entry => {
        visitContent[entry.path] = (await entry.content).toString()
      })
      expect(visitContent).toEqual(omit(['/some1'], fsData))
    })
  })

  describe('#overwrite', () => {
    it('throw error if target not exists', async () => {
      const tree = createTree({})
      await expect(
        tree.overwrite('/file/not/exists', 'new content'),
      ).rejects.toMatchSnapshot()
    })

    it('throw error if target is directory', async () => {
      const dirPath = '/some/path/to'
      const tree = createTree({ [dirPath + '/file']: 'content' })
      await expect(
        tree.overwrite(dirPath, 'new content'),
      ).rejects.toMatchSnapshot()
    })

    it('overwrite file content in OverlayHost', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree({ [filePath]: 'content' })
      const ofs = await getOverlayFs(tree)
      await tree.overwrite(filePath, 'content1')
      expect(ofs.readFileSync(filePath).toString()).toBe('content1')
      await tree.overwrite(filePath, Buffer.from('content2'))
      expect(ofs.readFileSync(filePath).toString()).toBe('content2')
    })

    it('overwrite file stat in OverlayHost', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree({ [filePath]: 'content' })
      const ofs = await getOverlayFs(tree)

      const fileModeBefore = ofs.statSync(filePath).mode
      expect(fileModeBefore & fs.constants.S_IRWXU).toBeTruthy()
      expect(fileModeBefore & fs.constants.S_IRGRP).toBeTruthy()
      expect(fileModeBefore & fs.constants.S_IROTH).toBeTruthy()

      await tree.overwrite(filePath, undefined, {
        mode: 0o733,
      })
      const fileModeAfter = ofs.statSync(filePath).mode
      expect(fileModeAfter & fs.constants.S_IFREG).toBeTruthy()
      expect(fileModeAfter & fs.constants.S_IRWXU).toBeTruthy()
      expect(fileModeAfter & fs.constants.S_IRGRP).toBeFalsy()
      expect(fileModeAfter & fs.constants.S_IROTH).toBeFalsy()
    })
  })

  describe('#create', () => {
    it('throw error if target already exists', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree({ [filePath]: 'content' })
      await expect(tree.create(filePath, 'content1')).rejects.toMatchSnapshot()
    })

    it('throw error if target is directory', async () => {
      const dirPath = '/some/path/to'
      const tree = createTree({ [dirPath + '/file']: 'content' })
      await expect(tree.create(dirPath, 'content')).rejects.toMatchSnapshot()
    })

    it('create file in OverlayHost', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree({ '/some/path/other/file': 'content1' })
      const ofs = await getOverlayFs(tree)
      await tree.create(filePath, 'content')
      expect(ofs.readFileSync(filePath).toString()).toBe('content')
      await tree.create(filePath + '1', Buffer.from('content'))
      expect(ofs.readFileSync(filePath + '1').toString()).toBe('content')
    })

    it('create file with stat in OverlayHost', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree({ '/some/path/other/file': 'content1' })
      const ofs = await getOverlayFs(tree)
      await tree.create(filePath, 'content', {
        mode: 0o733,
      })
      const fileMode = ofs.statSync(filePath).mode
      expect(fileMode & fs.constants.S_IFREG).toBeTruthy()
      expect(fileMode & fs.constants.S_IRWXU).toBeTruthy()
      expect(fileMode & fs.constants.S_IRGRP).toBeFalsy()
      expect(fileMode & fs.constants.S_IROTH).toBeFalsy()
    })
  })

  describe('#delete', () => {
    it('throw error if target is directory', async () => {
      const dirPath = '/some/path/to'
      const tree = createTree({ [dirPath + '/file']: 'content' })
      await expect(tree.delete(dirPath)).rejects.toMatchSnapshot()
    })

    it('throw error if target not exists', async () => {
      const tree = createTree({})
      await expect(tree.delete('/not/existed/path')).rejects.toMatchSnapshot()
    })

    it('delete file in OverlayHost', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree({ [filePath]: 'content' })
      const ofs = await getOverlayFs(tree)
      await tree.delete(filePath)
      expect(ofs.existsSync(filePath)).toBe(false)
    })
  })

  describe('#move', () => {
    it('throw error if source not exists', async () => {
      const tree = createTree({})
      await expect(
        tree.move('/not/existed/path', '/not/existed/path1'),
      ).rejects.toMatchSnapshot()
    })

    it('throw error if target is existed', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree({
        [`${filePath}1`]: 'content',
        [`${filePath}2`]: 'content',
      })
      await expect(
        tree.move(`${filePath}1`, `${filePath}2`),
      ).rejects.toMatchSnapshot()
    })

    it('move file from source to target in OverlayHost', async () => {
      const filePath = '/some/path/to/file'
      const tree = createTree({ [filePath]: 'content' })
      const ofs = await getOverlayFs(tree)
      await tree.move(filePath, `${filePath}1`)
      expect(ofs.existsSync(filePath)).toBe(false)
      expect(ofs.existsSync(filePath + '1')).toBe(true)
    })
  })
})

async function getOverlayFs(tree: Tree) {
  const host = await tree['_getHost']()
  return host['_fs']['fss'][0] as typeof fs
}

function createTree(fileStruct: { [path: string]: any } = {}) {
  return new Tree(() => new MemoryFsHost(Volume.fromJSON(fileStruct)))
}
