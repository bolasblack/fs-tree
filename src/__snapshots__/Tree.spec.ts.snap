// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree #create throw error if target already exists 1`] = `
FileAlreadyExistException(
  "Path \\"/some/path/to/file\\" already exist."
  { path: '/some/path/to/file' }
)
`;

exports[`Tree #create throw error if target is directory 1`] = `
FileAlreadyExistException(
  "Path \\"/some/path/to\\" already exist."
  { path: '/some/path/to' }
)
`;

exports[`Tree #delete throw error if target is directory 1`] = `
PathIsDirectoryException(
  "Path \\"/some/path/to\\" is a directory."
  { path: '/some/path/to' }
)
`;

exports[`Tree #delete throw error if target not exists 1`] = `
FileDoesNotExistException(
  "Path \\"/not/existed/path\\" does not exist."
  { path: '/not/existed/path' }
)
`;

exports[`Tree #get throw error if target is directory 1`] = `
PathIsDirectoryException(
  "Path \\"/some/path/to\\" is a directory."
  { path: '/some/path/to' }
)
`;

exports[`Tree #getDir throw error if target is file 1`] = `
PathIsFileException(
  "Path \\"/some/path/to/file\\" is a file."
  { path: '/some/path/to/file' }
)
`;

exports[`Tree #move throw error if source not exists 1`] = `
FileDoesNotExistException(
  "Path \\"/not/existed/path\\" does not exist."
  { path: '/not/existed/path' }
)
`;

exports[`Tree #move throw error if target is existed 1`] = `
FileAlreadyExistException(
  "Path \\"/some/path/to/file2\\" already exist."
  { path: '/some/path/to/file2' }
)
`;

exports[`Tree #overwrite throw error if target is directory 1`] = `
PathIsDirectoryException(
  "Path \\"/some/path/to\\" is a directory."
  { path: '/some/path/to' }
)
`;

exports[`Tree #overwrite throw error if target not exists 1`] = `
FileDoesNotExistException(
  "Path \\"/file/not/exists\\" does not exist."
  { path: '/file/not/exists' }
)
`;
